algorithms/C/arrays/mergingArray.c

```diff
@@ -0,0 +1,82 @@
+/******************************************************************************
+           Program to print the merged array without using third array
+******************************************************************************/
+
+#include<stdio.h>
+#define maxSize 100
+
+int main()
+{
+    int arrayOne[maxSize], sizeOne;
+    int arrayTwo[maxSize], sizeTwo;
+    int temp;
+
+    // Scanning the size of the array I
+    printf("Enter the size of the array I: ");
+    scanf("%d",&sizeOne);
+
+    // Scanning the array I of sizeOne
+    printf("Enter the %d elements of array I: \n",sizeOne);
+    for(int i=0; i<sizeOne; i++)
+    {
+        printf("Element [%d]: ",i);
+        scanf("%d",&arrayOne[i]);
+    }
+
+    // Scanning the size of the array II
+    printf("Enter the size of the array II: ");
+    scanf("%d",&sizeTwo);
+
+    // Scanning the array II of sizeTwo
+    printf("Enter the %d elements of array II: \n",sizeTwo);
+    for(int i=0; i<sizeTwo; i++)
+    {
+        printf("Element [%d]: ",i);
+        scanf("%d",&arrayTwo[i]);
+    }
+
+    // Assingning the last element of I array to the temp variable
+    temp = arrayOne[sizeOne];
+
+    // Size of array I is converted as follows
+    sizeOne += sizeTwo;
+
+    // Merging the array I and array II in array I
+    for(int i=0; i<sizeOne; i++)
+    {
+        if(arrayOne[i] == temp)
+        {
+            for(int j=0; j<sizeTwo; j++)
+            {
+                arrayOne[i++] = arrayTwo[j];      
+            }
+        }
+    }
```

Comment:
Right now the time complexity of this is `O(sizeOne)*O(sizeTwo)` which is Quadratic time. You can reduce the time complexity to linear time if you make a few simple changes, think about it and feel free to ask me if you need any help

Reply:
Thanks for your valuable review. Could you please help me in reducing its time complexity. It will be very helpful and grateful.

Reply:
First think about why you did 2 for loops.

You are using your first for loop to find the end of the `array1` and second for loop to insert elements of the `array2` to the end of the `array1`  basically merging it.

Here's the thing you don't need the first for loop because if all you need is to somehow have access to end of the array you can use the size of the array 

Let's assume `sizeOne = 3, sizeTwo = 3`  , so in a for loop if you set `i = sizeOne` you can start inserting the elements at the end of `array1` thus eliminating the need for the first for loop.
```C

      i = sizeOne = 3--|
                       |
                      \|/

 0     1     2         3            // index of the array          
-----------------------------------
 1  |  2  |  3  | (garbage value) | // array elements
-----------------------------------
```

Secondly to access the elements of the `array2`  you can try something like this
```C
arrayTwo[i-sizeOne]
```
Here you can't just directly use `i` because it's value is 3 and after index 2 there exist only garbage values, so to avoid that you have to subtract it to make it zero and since `i = sizeOne` it's resonable to use it to subtract.
```C
i = sizeOne = sizeTwo = 3, i < sizeOne+sizeTwo

i = 3, i - sizeOne = 0 // iteration-1
i = 4, i - sizeOne = 1 // iteration-2
i = 5, i - sizeOne = 2 // iteration-3

```
Hope this helps.